File: Problems/Trees/Implement-Trie-Prefix-Tree/Implement-Trie-Prefix-Tree.cs
using System;
using System.Collections.Generic;

// Using a Trie - Time O(w) for insert, search and startsWith, where w is the length of the word

public class TrieNode
{
    public TrieNode[] Next { get; } = new TrieNode[26];
    public bool Word { get; set; }
}

public class Trie
{
    private TrieNode root = new TrieNode();

    private TrieNode Find(string word)
    {
        var node = root;
        
        foreach (char c in word)
        {
            if (node.Next[c - 'a'] == null)
            {
                return null;
            }

            node = node.Next[c - 'a'];
        }

        return node;
    }

    public void Insert(string word)
    {
        var node = root;
        foreach (char c in word)
        {
            if (node.Next[c - 'a'] == null)
            {
                node.Next[c - 'a'] = new TrieNode();
            }

            node = node.Next[c - 'a'];
        }
        
        node.Word = true;
    }

    public bool Search(string word)
    {
        var node = Find(word);
        return node != null && node.Word;
    }

    public bool StartsWith(string prefix)
    {
        return Find(prefix) != null;
    }
}

class Program
{
    static List<bool> PerformOperations(List<string> operations, List<List<string>> inputs)
    {
        Trie obj = new Trie();
        List<bool> results = new List<bool>();

        for (int i = 0; i < operations.Count; i++)
        {
            if (operations[i] == "Trie")
            {
                obj = new Trie(); 
                results.Add(false); 
            }
            else if (operations[i] == "insert")
            {
                obj.Insert(inputs[i][0]);
                results.Add(false); 
            }
            else if (operations[i] == "search")
            {
                results.Add(obj.Search(inputs[i][0]));
            }
            else if (operations[i] == "startsWith")
            {
                results.Add(obj.StartsWith(inputs[i][0]));
            }
        }

        return results;
    }

    static void Main()
    {
        List<string> operations = new List<string> { "Trie", "insert", "search", "search", "startsWith", "insert", "search" };
        List<List<string>> inputs = new List<List<string>> { new List<string>(), new List<string> { "apple" }, new List<string> { "apple" }, new List<string> { "app" }, new List<string> { "app" }, new List<string> { "app" }, new List<string> { "app" } };

        List<bool> results = PerformOperations(operations, inputs);

        for (int i = 0; i < results.Count; i++)
        {
            Console.Write("Operation: " + operations[i]);
            if (inputs[i].Count > 0)
            {
                Console.Write(", Input: " + inputs[i][0]);
            }

            if (operations[i] == "Trie")
            {
                Console.WriteLine(", Result: Trie initialized.");
            }
            else if (operations[i] == "insert")
            {
                Console.WriteLine(", Result: null");
            }
            else if (operations[i] == "search")
            {
                Console.WriteLine(", Result: " + (results[i] ? "True" : "False"));
            }
            else if (operations[i] == "startsWith")
            {
                Console.WriteLine(", Result: " + (results[i] ? "True" : "False"));
            }
        }
    }
}

File: Problems/Trees/Implement-Trie-Prefix-Tree/Implement-Trie-Prefix-Tree.cpp
#include <iostream>
#include <string>
#include <vector>

// Using a Trie - Time O(w) for insert, search and startsWith, where w is the length of the word

struct TrieNode
{
    TrieNode *next[26] = {};
    bool word = false;
};

class Trie {
private:
    TrieNode root;

    TrieNode *find(std::string &word)
    {
        auto node = &root;
        for (char c : word)
        {
            if (!node->next[c - 'a'])
            {
                return nullptr;
            }

            node = node->next[c - 'a'];
        }
        
        return node;
    }

public:
    void insert(std::string word)
    {
        auto node = &root;
        for (char c : word)
        {
            if (!node->next[c - 'a'])
            {
                node->next[c - 'a'] = new TrieNode();
            }
            
            node = node->next[c - 'a'];
        }

        node->word = true;
    }

    bool search(std::string word)
    {
        auto node = find(word);
        return node && node->word;
    }

    bool startsWith(std::string prefix)
    {
        return find(prefix) != nullptr;
    }
};

std::vector<bool> performOperations(const std::vector<std::string> &operations, const std::vector<std::vector<std::string>> &inputs)
{
    Trie obj;
    std::vector<bool> results;

    for (size_t i = 0; i < operations.size(); ++i)
    {
        if (operations[i] == "Trie")
        {
            obj = Trie(); 
            results.push_back(false); 
        }
        else if (operations[i] == "insert")
        {
            obj.insert(inputs[i][0]);
            results.push_back(false); 
        }
        else if (operations[i] == "search")
        {
            results.push_back(obj.search(inputs[i][0]));
        }
        else if (operations[i] == "startsWith")
        {
            results.push_back(obj.startsWith(inputs[i][0]));
        }
    }

    return results;
}


int main()
{
    std::vector<std::string> operations = {"Trie", "insert", "search", "search", "startsWith", "insert", "search"};
    std::vector<std::vector<std::string>> inputs = {{}, {"apple"}, {"apple"}, {"app"}, {"app"}, {"app"}, {"app"}};

    std::vector<bool> results = performOperations(operations, inputs);

    for (size_t i = 0; i < results.size(); i++)
    {
        std::cout << "Operation: " << operations[i];
        if (!inputs[i].empty())
        {
            std::cout << ", Input: " << inputs[i][0];
        }

        if (operations[i] == "Trie")
        {
            std::cout << ", Result: Trie initialized." << std::endl;
        }
        else if (operations[i] == "insert")
        {
            std::cout << ", Result: null" << std::endl;
        }
        else if (operations[i] == "search")
        {
            std::cout << ", Result: " << (results[i] ? "True" : "False") << std::endl;
        }
        else if (operations[i] == "startsWith")
        {
            std::cout << ", Result: " << (results[i] ? "True" : "False") << std::endl;
        }
    }

    return 0;
}
